parsers/US_PREPA.py

```diff
@@ -0,0 +1,225 @@
+  res = r.get("https://aeepr.com/es-pr/generacion/Documents/Unidades_renovables.aspx")
+
+  assert res.status_code == 200, 'Exception when fetching renewable production for ' \
+                               '{}: error when calling url={}'.format(
+                                   zone_key, url)
+  print(res.text)
+  sourceData = extract_data(res.text)
+  print(sourceData)
+
+  original_renewable_output = renewable_output#If nothing gets subtracted renewable_output, there probably was no data on the renewables breakdown page
+
+  for item in sourceData:#Somewhat different from above, the item's label has the generation type and the item's value has generation in MW
+    if(item['label'] == "  "):#There's one empty item for some reason. Skip it.
+      continue
+
+    if(item['label'] == "Solar"):
+      data['production']['solar'] += float(item['value'])
+    elif(item['label'] == "Eolica"):
+      data['production']['wind'] += float(item['value'])
+    elif(item['label'] == "Landfill Gas"):
+      data['production']['biomass'] += float(item['value'])
+    else:
+      print("HEY! UNKNOWN RENEWABLE! " + item['label'])#TODO log warning if unknown renewable detected
+
+    renewable_output -= float(item['value'])#Subtract production accounted for from the renewable output total
+    print(renewable_output)
+
+    print(category)
+    print(outputInMW)
+
+
+  print(original_renewable_output)
+  print(renewable_output)
+
+  #Assume renewable generation not accounted for is hydro - if we could fetch the other renewable generation data
+  if(renewable_output > 0.0):#TODO this is negative sometimes, maybe add a log statement?
+    if(original_renewable_output == renewable_output):#Nothing got sutracted for Solar, Wind or Landfill gas - so the page probably didn't contain any data. Renewable type=unknown
+      data['production']['unknown'] += renewable_output
+    else:#Otherwise, any remaining renewable output is probably hydro
+      data['production']['hydro'] += renewable_output
+
+  #Step 3: fetch the timestamp, which is at the bottom of a different iframe
+  #Note: there's a race condition here when requesting data very close to <hour>:10 and <hour>:40, which is when the data gets updated
+  #Sometimes it's some seconds later, so we grab the timestamp from here to know the exact moment
+  res = r.get("https://aeepr.com/es-pr/generacion/Documents/CostosCombustible.aspx")#TODO do we know for sure the timestamp on this page gets updated *every time* the generation breakdown gets updated?
+
+  assert res.status_code == 200, 'Exception when fetching timestamp for ' \
+                               '{}: error when calling url={}'.format(
+                                   zone_key, url)
+
+
+  raw_timestamp = re.search(r"Ultima Actualización:  ((?:0[1-9]|1[0-2])/(?:[0-2][0-9]|3[0-2])/2[01][0-9]{2}  [0-2][0-9]:[0-5][0-9]:[0-5][0-9] [AP]M)", res.text).group(1)#Extract timestamp
```

Comment:
You can use beautiful soup and [arrow's tokens](https://arrow.readthedocs.io/en/latest/#supported-tokens) to make this much easier and more reliable.

Reply:
I already use arrow in `convert_timestamp`. Personally I think regular expressions are easier to read (and almost certainly use less memory than an HTML parser), so I kept this

Reply:
Correct on the memory usage but the bottleneck for this parser is likely in the network requests. I guess we'll have to disagree on the readability part.